### ISSUE TITLE
Bug: electrs 0.9.2 needs more time than 0.9.0

### ISSUE THREAD

**Issue by elkimek** [romanz/electrs#626]
**Describe the bug**

Follow up from: https://github.com/sparrowwallet/sparrow/issues/313

It seems to me that electrs 0.9.2 needs more time to communicate with other software components such as Dojo. I was on 0.9.0 where I had no issues using electrs but after upgrading to 0.9.2 I noticed timeout errors when importing xpub.

I tried to bump the timeout to 30s from default 10s and it helped. I had the exact same issue which I solved in the exact same way on two different implementations and hardware. Nodl (Rock64 with SATA) dockerless Dojo setup and on RoninDojo (RockPro64 with NVMe SSD). 

The file where I bumped the limit which helped with my timeout error (see to error in the link beneath): https://code.samourai.io/dojo/samourai-dojo/-/blob/develop/lib/indexer-rpc/rpc-client.js

**Electrs version**
0.9.2

**To Reproduce**
1. Use electrs with Dojo with default timeout limit 10000ms (see the file above)
2. Import xpub
3. Wait
4. See error*
5. Bump the limit
6. See if it helps

*https://i.postimg.cc/T37WL42m/photo-2021-11-11-08-47-24.jpg

**Additional context**
I also have huge issues syncing an account with long history from Sparrow using electrs. It's totally fine with smaller history accounts. From observing the logs I can tell that there's not much of activity as I assume there should be.

**Comment by Kixunil** [comment]
I think it'd be best to cut a new release (possibly with #543 for convenience) so Dojo/Nodl can update and we can check if the unreleased optimizations help. What do you think @romanz?

**Comment by romanz** [comment]
> What do you think @romanz?

Sounds good, will do later this week :)

**Comment by romanz** [comment]
-> #628

**Comment by romanz** [comment]
-> https://crates.io/crates/electrs/0.9.3 :)

**Comment by romanz** [comment]
https://github.com/sparrowwallet/sparrow/issues/313 is now merged.

**Comment by craigraw** [comment]
This is related - I am looking at introducing paging to the batched requests that Sparrow makes. Even though Sparrow increases the timeout for larger batches, this would effectively place a maximum on the number of ids in a particular batched request.

What would be a good default for the page size? I am thinking anywhere between 200 and 1000 at this stage.

**Comment by Kixunil** [comment]
I think it'd be best to test it rather than guess.

**Comment by craigraw** [comment]
Indeed, I have. But Electrum server implementations differ, as does their configuration, as does the hardware they run on. Just looking for suggestions here.

**Comment by Kixunil** [comment]
If you've found some correlation between configuration/hardware and performance I'd love to know. Maybe we can expose some parameters in RPC so that the wallets can adjust? Of course you can already adjust by detecting implementation.

**Comment by elkimek** [comment]
Maybe having some benchmark as proposed in https://github.com/romanz/electrs/issues/567 would help? Specifically the PTS/OBP platform would be awesome as it has great network effect and it's all open source.

**Comment by craigraw** [comment]
My thinking on it is this: Adding paging improves reliability, potentially at the cost of performance. Recent versions of Electrs approach or match ElectrumX in performance on typical (RPi4) hardware, even for queries over hundreds of ids. However, decreasing the page size (increasing the number of queries to load a wallet) can impact performance, particularly in a typical situation querying over Tor. Therefore, my goal was to find a point where the performance impact was minimal on a reasonably large wallet under these conditions.

This is an optimistic approach, in that I expect Electrum servers (and the node packages that contain them) will improve and solve performance issues over time. I don't want to by default limit wallet loading time to the lowest common denominator.

After much testing (mostly on a largish wallet with around 2000 addresses) I've settled on a default max page size of 500 ids in a batch request. Although it's probable that even larger wallets will notice some impact, I was unable to find a noticeable performance loss from larger page sizes on this wallet. On the other hand, decreasing the page size to 200 for example increased loading times by about 30%. 

That said, I've introduced the page size as a configurable option `"batchPageSize"` in the Sparrow config. @elkimek this may be a useful parameter to try with your larger wallets - it would be interesting to know if helps, either way.

Implemented in https://github.com/sparrowwallet/sparrow/commit/b013b5f50fbc7c12d7397b555b203bd33bb6357c.

**Comment by elkimek** [comment]
Thanks a lot, will try when new release comes out as I'm having some issues building it myself.

**Comment by Transisto** [comment]
Is 0.9 being tested against 0.8.9?

I don't have access to a 0.8.9 anymore but from tracking current 0.9.x speeds they're still at least 2x slower than 0.8.9
